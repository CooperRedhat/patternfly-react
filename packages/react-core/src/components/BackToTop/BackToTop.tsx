import * as React from 'react';
import styles from '@patternfly/react-styles/css/components/BackToTop/back-to-top';
import { css } from '@patternfly/react-styles';
import { Button } from '..';
import AngleUpIcon from '@patternfly/react-icons/dist/esm/icons/angle-up-icon';
import { canUseDOM } from '../../helpers/util';

interface BackToTopProps extends React.DetailedHTMLProps<React.HTMLProps<HTMLDivElement>, HTMLDivElement> {
  /** Additional classes added to the back to top. */
  className?: string;
  /** Title to appear in back to top button, defaults to 'Back to top'. */
  title?: string;
  /** Forwarded ref */
  innerRef?: React.Ref<any>;
  /** Selector for the scrollable element to spy on. Not passing a selector defaults to spying on window scroll events. */
  scrollableSelector?: string;
  /** Flag to always show back to top button, defaults to false. */
  alwaysShow?: boolean;
}

const BackToTopBase: React.FunctionComponent<BackToTopProps> = ({
  className,
  title = 'Back to top',
  innerRef,
  scrollableSelector,
  alwaysShow = false,
  ...props
}: BackToTopProps) => {
  const [hidden, setHidden] = React.useState(alwaysShow ? false : true);
  const [scrollElement, setScrollElement] = React.useState(null);

  const toggleHidden = () => {
    // scrollTop returns undefined in window use case
    const scrolled = scrollElement.scrollY ? scrollElement.scrollY : scrollElement.scrollTop;
    if (!alwaysShow) {
      if (scrolled > 400) {
        setHidden(false);
      } else {
        setHidden(true);
      }
    }
  };

  React.useEffect(() => {
    const hasScrollSpy = Boolean(scrollableSelector);
    if (hasScrollSpy) {
      const scrollEl = document.querySelector(scrollableSelector) as HTMLElement;
      if (!canUseDOM || !(scrollEl instanceof HTMLElement)) {
        return;
      }
      setScrollElement(scrollEl);
      scrollEl.addEventListener('scroll', toggleHidden);

      return () => {
        scrollEl.removeEventListener('scroll', toggleHidden);
      };
    } else {
      if (!canUseDOM) {
        return;
      }
      const scrollEl = window;
      setScrollElement(scrollEl);
      scrollEl.addEventListener('scroll', toggleHidden);

      return () => {
        scrollEl.removeEventListener('scroll', toggleHidden);
      };
    }
  });

  const handleClick = () => {
    scrollElement.scrollTo({ top: 0, behavior: 'smooth' });
  };

  return (
    <div
      className={css(styles.backToTop, hidden && styles.modifiers.hidden, className)}
      {...props}
      ref={innerRef}
      onClick={handleClick}
    >
      <Button component="a" variant="primary" icon={<AngleUpIcon aria-hidden="true" />} iconPosition="right">
        {title}
      </Button>
    </div>
  );
};

export const BackToTop = React.forwardRef((props: BackToTopProps, ref: React.Ref<any>) => (
  <BackToTopBase innerRef={ref} {...props} />
));
BackToTop.displayName = 'BackToTop';
