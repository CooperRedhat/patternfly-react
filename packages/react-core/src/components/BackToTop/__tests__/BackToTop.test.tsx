import React from 'react';
import { shallow } from 'enzyme';
import { BackToTop } from '../BackToTop';

test('BackToTop test', () => {
  const view = shallow(<BackToTop />);
  expect(view).toMatchSnapshot();
});
