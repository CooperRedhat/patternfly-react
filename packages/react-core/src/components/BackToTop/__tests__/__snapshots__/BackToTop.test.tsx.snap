// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BackToTop test 1`] = `
<BackToTopBase
  innerRef={null}
/>
`;
